### PR TITLE
[ENH] Conversational mode for GPT-3.5 models

### PR DESCRIPTION
**Pull request by paxcema**
## Description

**Fixes** #4892 

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📄 This change requires a documentation update

### What is the solution?

(Describe at a high level how the feature was implemented)

## Checklist:

- [ ] My code follows the style guidelines(PEP 8) of MindsDB.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation, or created issues to update them.
- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] I have shared a short loom video or screenshots demonstrating any new functionality.

### CODE REVIEW

File: mindsdb/integrations/handlers/openai_handler/openai_handler.py [hunk 1]
```diff
@@ -237,7 +250,7 @@ def predict(self, df, args=None):
 
         return pred_df
 
-    def _completion(self, model_name, prompts, api_key, api_args, args, parallel=True):
+    def _completion(self, model_name, prompts, api_key, api_args, args, df, parallel=True):
```
Comment (paxcema) [inline mindsdb/integrations/handlers/openai_handler/openai_handler.py:253]: remove `df`, unused

File: mindsdb/integrations/handlers/openai_handler/openai_handler.py [hunk 2]
```diff
@@ -303,8 +340,8 @@ def _tidy(comp):
                 max_batch_size = self.max_batch_size  # guards against changes in the API message
 
             # override batch_size to 1 if it's a conversational model:
-            if model_name in self.chat_completion_models:
-                max_batch_size = 1
+            # if model_name in self.chat_completion_models:
```
Comment (paxcema) [inline mindsdb/integrations/handlers/openai_handler/openai_handler.py:343]: remove leftover

File: mindsdb/integrations/handlers/openai_handler/openai_handler.py [hunk 3]
```diff
@@ -339,7 +376,45 @@ def _tidy(comp):
                 else:
                     completion['choices'].extend(p['choices'].result()['choices'])
 
-        return _tidy(completion)
+        return completion
+
+    def _truncate_msgs_for_token_limit(self, messages, model_name, max_tokens, truncate='first'):
```
Comment (paxcema) [inline mindsdb/integrations/handlers/openai_handler/openai_handler.py:381]: move into utilities

File: mindsdb/integrations/handlers/openai_handler/openai_handler.py [hunk 4]
```diff
@@ -339,7 +376,45 @@ def _tidy(comp):
...
+            if truncate == 'first':
+                messages = sys_priming + messages[2:]
+            else:
+                messages = sys_priming + messages[1:-1]
+
+            n_tokens = self._count_tokens(messages, encoder, model_name)
+        return messages
+
+    @staticmethod
+    def _count_tokens(messages, encoder, model_name='gpt-3.5-turbo-0301'):
```
Comment (paxcema) [inline mindsdb/integrations/handlers/openai_handler/openai_handler.py:404]: move into utilities

File: mindsdb/integrations/handlers/openai_handler/requirements.txt [hunk 5]
```diff
@@ -1,4 +1,5 @@
 mindsdb>=22.12.4.3
 mindsdb_sql >= 0.4.6, < 0.5.0
 pandas
-openai >= 0.27.0
\ No newline at end of file
+openai >= 0.27.0
+tiktoken >= 0.3.0
```
Comment (paxcema) [inline mindsdb/integrations/handlers/openai_handler/requirements.txt:5]: add this to gateway

Comment (paxcema) [reply mindsdb/integrations/handlers/openai_handler/requirements.txt:5]: https://github.com/mindsdb/mindsdb_gateway/pull/507

**Review by ZoranPandovski — APPROVED** [review]

**Review by ea-rus — APPROVED** [review]

**Pull request merged**